### PR TITLE
contrast-bias tool: debounce instead of reject based on time

### LINKED ISSUE #1359: Contrast/bias mouse control unusable
**Reporter: Pey-Lian Lim**

Some users are really stuck with this to the point they lose interest to try out other things.

Two problems:

1. Similar throttle problem as JDAT-2476.
2. The algorithm mapping mouse move to actual constrast/bias deltas need to be fine-tuned. What is DS9 doing that makes user so happy?

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request is an alternate to #1394 (although still needs to pull in the tests and perhaps contrast inversion from that PR) to optimize the drag behavior by replacing the rejection of events based on timestamp with debouncing in python (and caching the state and viewer size on dragstart).

Fixes #1359

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Review by pllim — COMMENTED** [review]
This seems way more responsive; in fact almost instantaneous on my machine! However, it does not update the contrast and bias sliders under Plot Options. Maybe we need some extra hooks into the sliders when using `debounced`? I don't see any errors in Python nor on browser dev console (the ones I see in dev console appear even on `main`, I think).

**Comment by pllim** [comment]
OMG... user error. I forgot Plot Options was on the image that I am not looking at. 😆  Sliders update fine. I think we should do it your way!

**Review by pllim — APPROVED** [review]

**Comment by kecnry** [comment]
> I think the buffer cannot catch up and I see some "blinks"  

This should be able to be adjusted with the `delay_seconds` on the debouncer.  In a perfect world, this would be machine and image-size dependent... but that would take some significant testing to find what relation is ideal.  But if catch ups are needed often, we should consider increasing it slightly.

**Comment by pllim** [comment]
No, the "blinks" don't happen enough to annoy me on my machine with example notebook. Another dev should try it though.

File: jdaviz/configs/imviz/plugins/tools.py [hunk 1]
```diff
@@ -134,59 +134,56 @@ class ContrastBias(CheckableTool):
...
-        # Note that we throttle this to 200ms here as changing the contrast
-        # and bias it expensive since it forces the whole image to be redrawn
-        if event == 'dragmove':
-            if (time.time() - self._time_last) <= 0.2:
-                return
+        # Normalize w.r.t. viewer display from 0 to 1
+        # Also see glue/viewers/matplotlib/qt/toolbar_mode.py
+        # bias range 0..1
+        # contrast range 0..4
+        bias = event_x / (self._max_x - 1)
```
Comment (rosteen) [inline jdaviz/configs/imviz/plugins/tools.py:160]: Huh, you know, I always thought this worked based on relative position to where you initially clicked, not compared to the edges of the viewer. No wonder I had a hard time using it 😆

File: jdaviz/configs/imviz/plugins/tools.py [hunk 2]
```diff
@@ -134,59 +134,56 @@ class ContrastBias(CheckableTool):
...
-            max_y = self.viewer.shape[0]
+        with delay_callback(self._layer_state, 'bias', 'contrast'):
+            self._layer_state.bias = bias
+            self._layer_state.contrast = contrast
 
-            if ((event_x < 0) or (event_x >= max_x) or
-                    (event_y < 0) or (event_y >= max_y)):
-                return
+    def on_mouse_or_key_event(self, data):
+        from jdaviz.configs.imviz.helper import get_top_layer_index
```
Comment (rosteen) [inline jdaviz/configs/imviz/plugins/tools.py:168]: What's the reason for this import to be here instead of at the top?

Comment (pllim) [reply jdaviz/configs/imviz/plugins/tools.py:168]: I think Kyle eventually wants to get rid of it altogether but I am not 100% sure either w.r.t. import move in this PR.

**Review by rosteen — APPROVED** [review]
Seems to work better than before - I had really given up on using it at all, I might actually use it now. Left one comment asking about the import location, but I'm approving on the assumption there's a good reasons that's not at the top of the file with everything else. I'll let @kecnry  merge though in case he does want to change it.

File: jdaviz/configs/imviz/plugins/tools.py (hunk 1 above)
Comment (pllim) [reply jdaviz/configs/imviz/plugins/tools.py:160]: Kyle did change the math to relative calculation but I made him change it back for this PR so we can compare performance on the mouse drag without worrying about the math.

If you think absolute calculation is the wrong way to do this, we can revisit in a future PR. Though with relative math, I am not sure how one can tell the mouse to drag to min or max in a consistent way. 🤷

**Comment by pllim** [comment]
Thanks, all!

**Pull request merged**